File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int K = sc.nextInt();
        int N = sc.nextInt();
        int[] arr = new int[K];
        long max = 0;
        
        for (int i = 0; i < K; i++) {
            arr[i] = sc.nextInt();
            max = Math.max(max, arr[i]);
        }
        sc.close();
        
        max++;
        long min = 0;
        long mid = 0;

        while (min < max) {
            mid = (max + min) / 2;
            long count = 0;
            for (int i = 0; i < arr.length; i++) {
                count += (arr[i] / mid);
            }
            if (count < N) {
                max = mid;
            } else {
                min = mid + 1;
            }
        }
        System.out.println(min - 1);
    }
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
K, N = map(int, input().split())
arr = [int(input()) for _ in range(K)]

max_val = max(arr) + 1
min_val = 0

while min_val < max_val:
    mid = (max_val + min_val) // 2
    count = sum([arr[i] // mid for i in range(len(arr))])
    if count < N:
        max_val = mid
    else:
        min_val = mid + 1
print(min_val - 1)